// better-auth/dist/integrations/react-start.cjs
'use strict';

require('../shared/better-auth.B6fIklBU.cjs');
require('@better-auth/utils/base64');
require('@better-auth/utils/hmac');
require('@better-auth/utils/binary');
const cookies_index = require('../cookies/index.cjs');
require('better-call');
require('zod/v4');
const socialProviders_index = require('../shared/better-auth.C-iQKHit.cjs');
require('../shared/better-auth.BIMq4RPW.cjs');
require('../shared/better-auth.CmL3zoTf.cjs');
require('../plugins/organization/access/index.cjs');
require('@better-auth/utils/random');
require('@better-auth/utils/hash');
require('@noble/ciphers/chacha');
require('@noble/ciphers/utils');
require('@noble/ciphers/webcrypto');
require('jose');
require('@noble/hashes/scrypt');
require('@better-auth/utils');
require('@better-auth/utils/hex');
require('@noble/hashes/utils');
require('../shared/better-auth.CYeOI8C-.cjs');
require('../shared/better-auth.CXhVNgXP.cjs');
require('kysely');
require('@better-auth/utils/otp');
require('../plugins/admin/access/index.cjs');
require('@better-fetch/fetch');
require('../shared/better-auth.DI0OyFsZ.cjs');
require('zod');
require('../plugins/custom-session/index.cjs');
require('@noble/hashes/sha3');
require('../shared/better-auth.ANpbi45u.cjs');
require('../shared/better-auth.C1hdVENX.cjs');
require('../shared/better-auth.D3mtHEZg.cjs');
require('../shared/better-auth.DRmln2Nr.cjs');
require('../crypto/index.cjs');
require('jose/errors');
require('defu');
require('../plugins/access/index.cjs');

const reactStartCookies = () => {
  return {
    id: "react-start-cookies",
    hooks: {
      after: [
        {
          matcher(ctx) {
            return true;
          },
          handler: socialProviders_index.createAuthMiddleware(async (ctx) => {
            const returned = ctx.context.responseHeaders;
            if ("_flag" in ctx && ctx._flag === "router") {
              return;
            }
            if (returned instanceof Headers) {
              const setCookies = returned?.get("set-cookie");
              if (!setCookies) return;
              const parsed = cookies_index.parseSetCookieHeader(setCookies);
              const { setCookie } = await import('../chunks/index.cjs');
              parsed.forEach((value, key) => {
                if (!key) return;
                const opts = {
                  sameSite: value.samesite,
                  secure: value.secure,
                  maxAge: value["max-age"],
                  httpOnly: value.httponly,
                  domain: value.domain,
                  path: value.path
                };
                try {
                  setCookie(key, decodeURIComponent(value.value), opts);
                } catch (e) {
                }
              });
              return;
            }
          })
        }
      ]
    }
  };
};

exports.reactStartCookies = reactStartCookies;
