// better-auth/dist/oauth2/index.mjs
export { c as createAuthorizationURL, f as decryptOAuthToken, e as encodeOAuthParameter, b as generateCodeChallenge, g as generateState, d as getOAuth2Tokens, h as handleOAuthUserInfo, p as parseState, r as refreshAccessToken, s as setTokenUtil, v as validateAuthorizationCode, a as validateToken } from '../shared/better-auth.BbL1R1Sz.mjs';
import 'zod/v4';
import 'better-call';
import '../shared/better-auth.CW6D9eSx.mjs';
import '@better-auth/utils/hash';
import '@better-auth/utils/base64';
import '../crypto/index.mjs';
import '@noble/ciphers/chacha';
import '@noble/ciphers/utils';
import '@noble/ciphers/webcrypto';
import 'jose';
import '@noble/hashes/scrypt';
import '@better-auth/utils';
import '@better-auth/utils/hex';
import '@noble/hashes/utils';
import '../shared/better-auth.B4Qoxdgc.mjs';
import '@better-auth/utils/random';
import '@better-fetch/fetch';
import '../shared/better-auth.CuS_eDdK.mjs';
import '../shared/better-auth.CMQ3rA-I.mjs';
import '../shared/better-auth.DdzSJf-n.mjs';
import '../cookies/index.mjs';
import '@better-auth/utils/hmac';
import '../shared/better-auth.tB5eU6EY.mjs';
import '@better-auth/utils/binary';
import 'jose/errors';
import '../shared/better-auth.DBGfIDnh.mjs';
import '../shared/better-auth.n2KFGwjY.mjs';
import 'defu';
